src/Transports.AspNetCore/HttpPostValidationRule.cs

```diff
@@ -0,0 +1,23 @@
+#nullable enable
+
+using GraphQL.Server.Transports.AspNetCore.Errors;
+using GraphQL.Validation;
+using GraphQLParser.AST;
+
+namespace GraphQL.Server.Transports.AspNetCore;
+
+/// <summary>
+/// Validates that HTTP POST requests do not execute subscriptions.
+/// </summary>
+public sealed class HttpPostValidationRule : IValidationRule
+{
+    /// <inheritdoc/>
+    public ValueTask<INodeVisitor?> ValidateAsync(ValidationContext context)
+    {
+        if (context.Operation.Operation == OperationType.Subscription)
+        {
+            context.ReportError(new HttpMethodValidationError(context.Document.Source, context.Operation, "Subscription operations are not supported for POST requests."));
```

Comment:
And yet I do not understand. GraphQL client can not send a subscription request through the POST? Should the client upgrade HTTP connection to WebSockets one and then communicate with server sending subscription?

Reply:
Yes, that is how it must be done.  It is impossible to send a subscription request via POST.  Once upgraded to a WebSocket connection, the WebSocket handler takes over.  During execution within the WebSocket handler, execution will not include this validation rule, of course, as request over WebSockets supports queries, mutations, or subscription requests.

If you're still a bit confused, I can write additional detail.  It is not hard to understand, but it would take me a while to type it out in a well-written form.

Reply:
>  **It is impossible** to send a subscription request via POST. 

😮  I'm just want to undestand whether you are talking about some _ideal_ implementation that conforms to spec or about current server v6 implementation. I suppose the former because I'm 100% sure that we use _regular_ HTTP POST for queries, mutations **and** subscriptions.   

I agree that I **should** read more to be familiar with all these concepts (sorry, no much free time). Now I rather want to place accents, especially about what is possible and what impossible. Otherwise, I begin to make the impression that 2 + 2 = 5.

Reply:
Ok, I will try to explain.  When a POST request comes in, this happens: (generally)

1. The request is parsed
2. The document is validated
3. The document is executed
4. The response is serialized
5. The response is written to the output
6. The connection is closed

If a subscription request came in over a POST connection, what would happen?

1. _The request is parsed_.  Ok.
2. _The document is validated_.  Ok.
3. _The document is executed_.   What does "executed" mean?  It means returning an `IObservable<object?>` (aka event stream) to the middleware, according to spec.  There is no actual 'execution' until an event occurs.
4. _The response is serialized_.  What response?  There is none.  Until an event occurs, there is no response to write.  There is no "success" response.  There is nothing to write, according to the GraphQL spec; not even an acknowledgement.
5. _The response is written to the output_.  So let's say we return a generic "success" response here, similar to a WebSocket acknowledgement message.
6. _The response is closed_.  Ok, so the response is closed -- and the event stream is closed/stopped.

Essentially, in an 'ideal' world, the observable would be subscribed to and almost immediately disposed, leaving a few mere milliseconds between attaching to the event stream and disconnecting from it.

Ok let's say that was possible.  Let's say 3 events arrived in those few milliseconds.  Where would they be written?  Each one needs to be written as a complete response message, according to spec.  And there's the connection acknowledgement message.  Do we write them all sequentially to the output?  As a JSON list?  Multipart MIME message?

As you can see, it just doesn't make sense for a POST request to return a subscription.  Certainly there is no code to support it now.  If you attempted to do so, the returned `IObservable<object>` would never be subscribed to and the response would appear to be empty (because the `SubscriptionDocumentExecuter` does not set `Data`).

Now it is certainly possible to write other protocols, which may even work over POST.  For example, a "long polling" approach (often used in the earlier days of the web for "push" notifications) could be used whereby the POST connection stays open without writing anything to the output until a data event is written.  However, specifications would need to be written to say how lunch the connection stays open, what happens in an idle connection, can multiple messages be sent, etc.  Our server might need to keep the subscription alive while clients reconnect, queuing events.  In any case, that protocol does not currently exist for GraphQL.  Rather, there are three WebSockets-based protocols which provide that functionality.  They provides all the specifications for how to set up the connection, return results, issue additional queries over the same connection, and so on.  The oldest one nobody uses at all.  The older one is what is supported by the Server project now.  The newest one is what is currently recommended by Apollo and is not yet supported by the Server project.

So when I say that it is impossible to execute a subscription over a POST connection, I'm saying that there are no protocols existing right now for a server or client to adhere to that would specify the mechanism by which a subscription could take place over POST.  And it certainly cannot take place over the existing protocols.  So essentially, it's impossible now.  Yes, someone can invent their own protocol to do so, but with WebSocket support available now, I feel it is unlikely.

Reply:
> I'm 100% sure that we use regular HTTP POST for queries, mutations and subscriptions.

No, it's literally impossible with the existing code.  Yes, the server will return an empty GraphQL result, but it is meaningless and the `IObservable<object?>` is never subscribed to.  All _working_ subscription requests comes in over a WebSocket connection.

Now may be WebSocket connections come in as POST connections and get upgraded to WebSockets.  I don't know; ASP.NET Core hides that from the middleware.  But as distinct from GET and POST, you simply cannot execute a subscription - with any results - via POST with the v6 code.  I didn't test it, but I'm pretty sure you will always get something like this (unless an error occurs):

```json
{
    "data": null
}
```

There's just no code that could allow for anything different.  `Data` is simply never set, and `Streams` is never read.  See:
- https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Execution/SubscriptionExecutionStrategy.cs#L57

Reply:
Here is a protocol for subscriptions over "Server-Sent Events" which I think is a long-lived HTTP POST connection.

https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md

Both the current server project and my updated code do not support this protocol.

Reply:
As a side note here, I believe that WebSocket requests actually arrive as GET requests, along with special headers that identify the request as a WebSocket connection request.